File: Dynamic_programming/practice_1.cpp
#https:  // www.acmicpc.net/problem/1932

#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>

int main() {
    int n;
    std::cin >> n;
    int table[2][n];
    std::cin.ignore();
    for (int i = 0; i < n; ++i) {
        std::string S, T;
        int prev = (i - 1) % 2, cur = i % 2;
        getline(std::cin, S);
        std::stringstream X(S);
        int each_idx = 0;
        while (getline(X, T, ' ')) table[cur][each_idx++] = stoi(T);

        if (i == 0) continue;

        for (int j = 0; j < i + 1; ++j) {
            int left = j > 0 ? table[prev][j - 1] : 0;
            int right = j < i ? table[prev][j] : 0;
            table[cur][j] = table[cur][j] + (left > right ? left : right);
        }
    }
    int f_pos = (n - 1) % 2;
    std::cout << *std::max_element(table[f_pos], table[f_pos] + n) << std::endl;
}

File: Dynamic-programming/practice_1.py
# https://www.acmicpc.net/problem/1932

n = int(input())
table = []
table.append([int(input())])
for i in range(1, n):
    table.append(list(map(int, input().split())))
    for j, e in enumerate(table[i]):
        best = 0
        if j > 0:
            best = table[i - 1][j - 1]
        if j < len(table[i]) - 1:
            best = max(best, table[i - 1][j])
        table[i][j] = e + best
print(max(table[n - 1]))
